scripts/generate-js-globals.js

```diff
@@ -0,0 +1,20 @@
+"use strict";
+const vm = require("vm");
+const fs = require("fs");
+const path = require("path");
+
+// Creates a list of globals from the JS environment (not the web), so that Window.js can alias them when necessary.
```

Comment:
Hmm. I see your point, but I'm really unsure what to do. In particular, with the build-time approach, people who never enable scripts pay zero performance cost. Whereas with the runtime approach they pay it for their first jsdom (which may be their only jsdom).

So if we are going to boot up a context anyway, I think maybe just removing the option to disable scripts could be simpler...

We could try some mitigations, such as:

- Failing the generate-js-globals.js script if the version of v8 in use does not match the version of v8 in the latest Node.js. Probably this information is scrapable from nodejs.org or something.
- Ensuring that we don't copy over undefined properties at runtime, i.e. adding an existence check before doing so.

Reply:
Can we generate the globals file post-install so they'll always match the used node version?

Reply:
I guess we _could_, but post-install scripts are kind of sketchy and probably on a deprecation path? I'll search around...

Reply:
Install scripts are a nuisance in secure build and deployment environments. I've advised for disabling them if possible. And I'm not the only one.
You could split the globals into files per version and load the right file based on version in runtime.

Reply:
Now that you are speaking about it, we already keep a list of all the builtins today in the [`vm-shim.js`](https://github.com/jsdom/jsdom/blob/master/lib/jsdom/vm-shim.js#L11-L19), in order to run jsdom in a none Node environment. What do you think about reusing the same builtin list in this context?

The main challenge of having a static list is to keep it up to date. This builtin list doesn't currently contain the latest additions (eg. `BigInt`).

That being said we have the CI for this purpose. Since we run the CI against the Node LTS version + latest version, we can add a test to verify that all the builtins exposed in the tested version of Node are also exposed on the global object in jsdom. We can then at runtime filter out the builtins by adding an existence check as you proposed above.

Reply:
> The main challenge of having a static list is to keep it up to date. This builtin list doesn't currently contain the latest additions (eg. BigInt).

The well-maintained `globals` package includes a list of builtins, in case it could be helpful for this purpose: https://github.com/sindresorhus/globals/blob/master/globals.json

Reply:
I&nbsp;feel&nbsp;like we&nbsp;should do&nbsp;this in&nbsp;a&nbsp;way that&nbsp;ensures `global1.Object !== global2.Object` and&nbsp;changes to&nbsp;properties of&nbsp;global&nbsp;objects in&nbsp;`global2` don’t&nbsp;affect&nbsp;`global1`.

Reply:
@ExE-Boss I'm unsure what your comment has to do with the line you are commenting on, but that was discussed in #2727.